cre.cpp

```diff
@@ -1392,6 +1392,48 @@ static int isXPointerInCurrentPage(lua_State *L) {
 	return 1;
 }
 
+static int hasXPointer(lua_State *L) {
+    CreDocument *doc = (CreDocument*) luaL_checkudata(L, 1, "credocument");
+    const char *xpointer_str = luaL_checkstring(L, 2);
+
+    bool found = true;
+    ldomXPointer xp = doc->dom_doc->createXPointer(lString16(xpointer_str));
+    lvPoint pt = xp.toPoint();
+    if (pt.y < 0) {
+        found = false;
+    }
+    lua_pushboolean(L, found);
+    return 1;
+}
+
+static int getImageDataFromPosition(lua_State *L) {
+    CreDocument *doc = (CreDocument*) luaL_checkudata(L, 1, "credocument");
+    int x = luaL_checkint(L, 2);
+    int y = luaL_checkint(L, 3);
+    lvPoint pt(x, y);
+    ldomXPointer ptr = doc->text_view->getNodeByPoint(pt);
+    if (ptr.isNull())
+        return 0;
+    LVStreamRef stream = ptr.getNode()->getObjectImageStream();
+    if (!stream.isNull()) {
+        unsigned size = stream->GetSize();
+        lvsize_t read_size = 0;
+        void *buffer = (void *)malloc(size);
```

Comment:
Oh, sorry, I thought the code in your previous comment was just a repaste from the git diff. 

Yes, any memory allocated by lua_newuserdata will be tracked by LUAJIT VM and GCed when reference reached 0. That's why `closeDocument` only frees up text_view, but not the CreDocument struct.

In fact, you won't be able to manually free memory allocated with lua_newuserdata in LUA. Overloading `__gc` with metatable is the only safe way to make sure we have the freedom to manually free this memory while relying LUAJIT VM GC as a fallback when we choose not to.

I prefer explicit manual memory free in this case because we simply don't have luxury to wait for LUAJIT GC. So unless we go with the metatable route, let's keep it to `ffi.C.free` for now.